lisa/tests/staging/scheduler/util_tracking.py

```diff
@@ -0,0 +1,338 @@
+
+        The integral of `util_est_enqueued` is expected to be always not
+        smaller than that of `util_avg`, since this last is subject to decays
+        while the first not.
+
+        On fast-ramp systems, the `util_est_ewma` signal is never smaller then
+        the `util_est_enqueued`, thus his integral is expected to be bigger.
+
+        On non fast-ramp systems instead, the `util_est_ewma` is expected to be
+        smaller then `util_est_enqueued` in ramp-up phases, or bigger in
+        ramp-down phases.
+
+        Those conditions are checked on a single execution of a task which has
+        three main behaviours:
+        - STABLE: periodic big task running for a relatively long period to
+                  ensure `util_avg` saturation.
+        - DOWN: periodic ramp-down task, to slowly decay `util_avg`
+        - UP: periodic ramp-up task, to slowly increase `util_avg`
+
+        """
+        failure_reasons = {}
+        metrics = {}
+
+        # We have only two task: the main 'rt-app' task and our 'test_task'
+        test_task = self.trace.analysis.rta.rtapp_tasks[-1]
+
+        ue_df = self.trace.df_events('sched_util_est_task')
+        ue_df = ue_df[ue_df.pid == test_task.pid]
+        ua_df = self.trace.df_events('sched_load_se')
+        ua_df = ua_df[ua_df.pid == test_task.pid]
+
+        failures = []
+        for phase in self.trace.analysis.rta.task_phase_windows(test_task):
+            phase_df = ue_df[phase.start:phase.end]
+            area_enqueued = series_integrate(phase_df.util_est_enqueued)
+            area_ewma = series_integrate(phase_df.util_est_ewma)
+
+            phase_df = ua_df[phase.start:phase.end]
+            area_util = series_integrate(phase_df.util)
+
+            metrics[phase.id] = self.PhaseStats(phase.start, phase.end,
+                                                area_util, area_enqueued, area_ewma)
+
+            phase_name = "phase {}".format(phase.id)
+            if area_enqueued < area_util:
+                failure_reasons[phase_name] = 'Enqueued smaller then Util Average'
+                failures.append(phase.start)
+                continue
+
+            # Running on FastRamp kernels:
+            if self.fast_ramp:
+
+                # STABLE, DOWN and UP:
+                if area_ewma < area_enqueued:
+                    failure_reasons[phase_name] = 'NO_FAST_RAMP: EWMA smaller then Enqueued'
+                    failures.append(phase.start)
+                    continue
+
+            # Running on (legacy) non FastRamp kernels:
+            else:
+
+                # STABLE: ewma ramping up
+                if phase.id == 0 and area_ewma > area_enqueued:
+                    failure_reasons[phase_name] = 'FAST_RAMP(STABLE): EWMA bigger then Enqueued'
+                    failures.append(phase.start)
+                    continue
+
+                # DOWN: ewma ramping down
+                if 0 < phase.id < 5 and area_ewma < area_enqueued:
```

Comment:
Fiddling with phase ids is finicky, we should do something like allow specifying a friendly name to an `RTATask` and have it copied in its phases. That way we could do something like
`if rtapp_task.phases[phase.id].name == "ramp_down" and ...`

That way we can change the rtapp profile without having to figure out how many phases it generates. It's not a problem here because we pin the task on a big CPU, but if we change that in the future we'll need to adapt the profile with the capacity landscape, which might result in a dynamically varying number of phases anyway...

Reply:
Are you handling that as part of the wlgen update you are baking atm ?

Reply:
Not yet, but can/should do.